kotlin/workflow-core/src/main/java/com/squareup/workflow/WorkflowAction.kt

```diff
@@ -13,33 +13,52 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
+@file:Suppress("DeprecatedCallableAddReplaceWith")
+
 package com.squareup.workflow
 
+import com.squareup.workflow.WorkflowAction.Mutator
+
 /**
  * A function that can change the current state of a [Workflow] by returning a new one, and
  * also optionally emit an output.
  */
-interface WorkflowAction<StateT, out OutputT : Any> : (StateT) -> Pair<StateT, OutputT?> {
+interface WorkflowAction<StateT, out OutputT : Any> {
+  class Mutator<S>(var state: S)
 
-  override operator fun invoke(state: StateT): Pair<StateT, OutputT?>
+  /**
+   * Called with the current state of a workflow, allowing the state to be updated.
+   *
+   * @return an output value for the workflow to emit, or null to emit nothing.
+   */
+  fun Mutator<StateT>.apply(): OutputT?
 
   companion object {
 
     /**
-     * Creates a [WorkflowAction] from the [update] lambda.
+     * Convenience to create a [WorkflowAction] from the [apply] lambda.
```

Comment:
@davidapgar @bencochran @timdonnelly I don't believe that Swift has an analog to this, and I'm curious to hear your reactions to it. 

I've been pretty happy with how it rolled out across the Kotlin demos (I'll highlight a few). It allows us some flexibility when full fledged hierarchy of action types are more trouble than they're worth, especially in spots where it allowed discrete action functions to get the job done more concisely.

Reply:
If I'm understanding this correctly, we do have an equivalent in Swift:
https://github.com/square/workflow/blob/master/swift/Workflow/Sources/WorkflowAction.swift#L42

Specifically, it's a constructor to create an `AnyWorkflowAction` (type erased) with a lambda for the `apply`, instead of having to make a full type.

Not something I've seen used in practice (aside from in a few test cases in the infrastructure), but I can see the convenience.

Reply:
Oh, nice! I totally missed that.